modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php

```diff
@@ -0,0 +1,259 @@
+   */
+  protected $entityTypeManager;
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\migrate\MigrateException
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    ConfigManagerInterface $config_manager,
+    EntityTypeManagerInterface $entity_type_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    // The default configuration.
+    $this->configuration += [
+      'workflow_config_name' => 'workflows.workflow.oe_corporate_workflow',
+      'published_state' => 'published',
+      'unpublished_state' => 'draft',
+    ];
+    $this->configManager = $config_manager;
+    $this->entityTypeManager = $entity_type_manager;
+    // Check if the configuration is set correctly.
+    $this->validateConfigurationKeys();
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\migrate\MigrateException
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition, MigrationInterface $migration = NULL) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('config.manager'),
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\migrate\MigrateException
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+    // The expected value has to be a string but the source entity might not
+    // have a moderation state defined, in that case stringify the NULL value
+    // and continue.
+    $value = is_null($value) ? (string) $value : $value;
+    // Check if the source value is of the expected type.
+    if (!is_string($value)) {
+      throw new MigrateException(sprintf('%s is not a string.', var_export($value, TRUE)));
+    }
+    // Check whether the current workflow state is a valid one or not. If it's
+    // not a valid one, set it according to the entity "status". In this case,
+    // if the status is set to "1", set it to the given "published_state", in
+    // any other case, set it to the given "unpublished_state".
+    $workflow = $this->getWorkflow($this->configuration['workflow_config_name']);
+    if ($this->isValidWorkflowState($workflow, $value) === FALSE) {
+      $value = (int) $row->getSourceProperty('status') === 1 ? $this->configuration['published_state'] : $this->configuration['unpublished_state'];
```

Comment:
Why are you hardcoding the `status` field in here? Make this better configurable. What if I have a migration where that value is under a `published` key. You can move this up to the plugin config and default it to `status` if you want.

Reply:
I'm not sure about it. Do we need to make this field configurable? I cannot find an use case in which someone need to manage the entity status with a different field. BTW, I added it as a configuration value and use it with default value as "status".

Reply:
I'm sorry but I'm missing the point of this comment.
The plugin, in its initial state, provides three functionalities
- checks if the input value is a valid state
- if not a valid state, defaults to the entity status
- if valid, verifies that the entity status and the state status are in accordance
By rendering the "status" configurable, we lose two of the functionalities.

Also
- I cannot see which other field could be used as a fallback value to set the workflow state when the provided value is NULL or not a valid one
- we open the "doors" to faulty configuration, e.g. users are able to use fields like `title` or `body/0/value`, which doesn't make sense
- we might have to deal with corrupted data and now we're losing the benefit to verify if entity status and workflow state status are in accordance. if we remove the accordance validation, we might end with nodes that have "status = 1" and "moderation_state = draft"

@samuelsolis , when you refactor code, could you please ensure that the class description, the comments and the code are talking about the same thing.
Comparing a typecasted input value to int, to the state status doesn't make any sense, it looks more like a bug, could you please remove it?

Thanks.

Reply:
Yes, you're right although I'm not sure exactly the meaning of "status" in this contexts. I'm waiting for a Daniel answer , but for sure I can remove this check and check again the comments to ensure that it's correct.

Reply:
> I'm not sure about it. Do we need to make this field configurable? I cannot find an use case in which someone need to manage the entity status with a different field. 

You are looking at this purely from the perspective of migrating froma D7 database. This component can be used for migrating from wherever. If I have a CSV file where the "status" field is called published, I should be able to still use it the same way.

> By rendering the "status" configurable, we lose two of the functionalities.

Absolutely not true. The fact that we can specify the name of the field should have no impact in the provided benefits of it. I asked to make only the field *name* configurable. 

> I cannot see which other field could be used as a fallback value to set the workflow state when the provided value is NULL or not a valid one

No other field would be used, but the field NAME can be different. Don't think only at D7 - D8 migrations.

> we open the "doors" to faulty configuration, e.g. users are able to use fields like title or body/0/value, which doesn't make sense

yes, it doesnt make sense. But there aren't simple "users" who create migrations. Whoever uses this component is responsible for writing correct migrations. You, for example, will use the "status" field. What users are you afraid of?